File: src/character.py
from serialize import Serializable
import copy

class Character(Serializable):
    uid = ord('a')

    def __init__(self, teampos : int, name : str, maxHP : int, softAtk : int, hardAtk : int, mov : int, rng : int, netWorth : int, arcAtk : bool = True, flavor : str = ''):
        self.name = name
        self.teampos = teampos
        self.team = None
        self.maxHP = maxHP
        self.HP = maxHP
        self.defShieldHP = 0
        self.softAtk = softAtk
        self.hardAtk = hardAtk
        self.mov = mov
        self.rng = rng
        self.arcAtk = arcAtk
        self.lastMoved = 0
        self.position = [None, None]
        self.netWorth = netWorth
        self.flavor = flavor
        self.cid = chr(Character.uid)
        (self.turnMoved, self.turnAttacked) = (None, None)
        Character.uid += 1
    def copy(self):
        return copy.copy(self)
    def beginTurn(self):
        me = self
        if self.defShieldHP > 0:
            me = self.copy()
            me.HP += 50
            me.defShieldHP = 0
        return me
    def takeDmg(self, isHard, power):
        if self.defShieldHP > 0:
            shielded = min(self.defShieldHP, power)
            self.defShieldHP -= shielded
            power -= shielded
        self.HP -= power
        return power
    def getAtk(self, isHard, turnId):
        if self.turnAttacked == turnId - 1: # "attacked last turn"
            return 10
        elif isHard:
            return self.hardAtk
        else:
            return self.softAtk
    def buff(self):
        self.defShieldHP = self.maxHP  #min(self.maxHP, self.HP + 50)
        return self.defShieldHP

    def serialize(self):
        return {
            "name" : self.name,
            "cid" : self.cid,
            "maxHP" : self.maxHP,
            "HP" : self.HP,
            "softAtk" : self.softAtk,
            "hardAtk" : self.hardAtk,
            "mov" : self.mov,
            "rng" : self.rng,
            "netWorth" : self.netWorth,
            "flavor" : self.flavor
        }

File: src/character.cpp
#include "character.hpp"
#include "nlohmann/json.hpp"

using json = nlohmann::json;
char Character::s_uid = 'a';

Character::Character(const ImmutableCharacter& im):
    uid(s_uid++),
    im(im), HP(im.maxHP), pos(0, 0),
    turnMoved(-1), turnAttacked(-1),
    defShieldHP(0)
{
};

ImmutableCharacter::ImmutableCharacter(const char* name, short maxHP, short softAtk, short hardAtk, uint8_t mov,
                     uint8_t rng, float netWorth, bool usesArcAttack, const char* flavor) : usesArcAttack(usesArcAttack) {
    this->name = name;
    this->maxHP = maxHP;
    this->softAtk = softAtk;
    this->hardAtk = hardAtk;
    this->mov = mov;
    this->rng = rng;
    this->netWorth = netWorth;
    this->flavor = flavor;
    this->maxAtk = std::max(softAtk, hardAtk);
}

ImmutableCharacter::~ImmutableCharacter(){
    for(auto effect_ptr : specialAction)
        delete effect_ptr;
}

inline bool rndBool(Generator& gen){ return (gen() >> 6) % 2; } //apparently, the lowest bit is often not that random, so I take the 7th one

ImmutableCharacter ImmutableCharacter::random(Generator& gen){
    return ImmutableCharacter("Bob",
        std::uniform_int_distribution<short>(0, 150)(gen),
        std::uniform_int_distribution<short>(0, 80)(gen),
        std::uniform_int_distribution<short>(0, 80)(gen),
        std::uniform_int_distribution<unsigned char>(0, 5)(gen),
        std::uniform_int_distribution<unsigned char>(0, 5)(gen),
        0,
        rndBool(gen)
    );
}

short Character::takeDmg(bool isHard, short power) {
    (void) isHard; //currently no difference between hard and soft damage
    if(this->defShieldHP > 0) {
        short shielded = std::min(this->defShieldHP, power);
        this->defShieldHP -= shielded;
        power -= shielded;
    }
    this->HP -= power;
    return power;
}

short Character::getAtk(bool isHard, short turnID) const {
    if(this->turnAttacked == turnID - 1) {
        return 10;
    } else if(isHard) {
        return im.hardAtk;
    } else {
        return im.softAtk;
    }
}

void Character::buff() {
    this->defShieldHP = im.maxHP;
}
